specs/Tranche.spec.sol

```diff
@@ -0,0 +1,235 @@
+        num redemptionAmount = amount * yieldRate(SELL);
+
+        require(it[owner] >= amount);
+        it[owner] -= amount; // no underflow
+
+        num share = position.withdraw(redemptionAmount);
+        // assert(vault.sellPrice() == redemptionAmount / share);
+        token.transferFrom(this, owner, redemptionAmount);
+    }
+
+    // @inv totalValue(BUY) >= totalValue(SELL)
+    function totalValue(mode) view {
+        return position.value(this, mode);
+    }
+
+    // yield per interest token
+    // @inv yieldRate(BUY) >= yieldRate(SELL)
+    function yieldRate(mode) view {
+        if (hasLoss(mode)) return 0;
+
+        num currentYield = totalValue(mode) - sum(pt); // no underflow
+        return currentYield / sum(it); // NOTE: this could be bigger than 1, if the value becomes more than double.
+    }
+
+    // loss in principal
+    // @inv hasLoss(BUY) implies hasLoss(SELL)
+    function hasLoss(mode) view {
+        return totalValue(mode) < sum(pt);
+    }
+}
+
+spec Position {
+    Vault vault; // many positions can be associated with the same vault
+    Token token;
+
+    mapping(address => num) shares; // position shares // code.WrappedPosition.balanceOf
+
+    num reserveUnderlying;
+    num reserveShares;
+    mapping(address => num) reserveBalances;
+
+    /*
+     * @inv token.balanceOf(this) >= reserveUnderlying
+     * @inv vault.shares[this] == reserveShares + sum(shares)
+     */
+
+    function value(address owner, mode) view {
+        num price = vault.sharePrice();
+        if (mode == BUY) price = vault.buyPrice();
+        if (mode == SELL) price = vault.sellPrice();
+        return shares[owner] * price;
+    }
+
+    // deposit underlying
+    function deposit(num amount) {
+        num share = amount / vault.buyPrice();
+        token.transferFrom(msg.sender, this, amount);
+
+        /* no reserve version:
+        num cost = vault.buy(share); // assert(cost == amount);
+        */
+        reserveUnderlying += amount;
+        if (reserveShares < share) buyReserve();
+        // assert(reserveShares >= share);
+        reserveShares -= share; // no underflow
+
+        shares[msg.sender] += share;
+        return share;
+    }
+
+    // withdraw underlying
+    function withdraw(num amount) { // code.WrappedPosition.withdrawUnderlying()
+        num share = amount / vault.sellPrice();
+        require(shares[msg.sender] >= share);
+        shares[msg.sender] -= share; // no underflow
+
+        /* no reserve version:
+        num proceeds = vault.sell(share); // assert(proceeds == amount);
+        */
+        reserveShares += share;
+        if (reserveUnderlying < amount) sellReserve();
+        // assert(reserveUnderlying >= amount);
+        reserveUnderlying -= amount; // no underflow
+
+        token.transferFrom(this, msg.sender, amount);
+        return share;
+    }
+
+    // @inv old(token.balanceOf(this) - reserveUnderlying) == new(token.balanceOf(this) - reserveUnderlying)
```

Comment:
We can add more properties, which would be great.   I guess what you meant was `old(token.balanceOf(this)) == (new(vault.balanceOf(this)) - old(vault.balanceOf(this))) * buyPrice()`, right?  But, please note that it is possible to have `token.balanceOf(this) > reserveUnderlying`, because someone can "donate" tokens to a position contract.